### ISSUE TITLE
CP function

### ISSUE THREAD

**Issue by JeanKossaifi** [tensorly/tensorly#470]
CP via ALS is probably the most used function in TensorLy and comes with lots of options. One issue is that due to these successive additions, bugs (see e.g. this [commit](d66110f7c961ce896a051b446a23d69bd54ecc8e)) and undue complexity are slowly creeping in while the code is becoming increasingly hard to read.

Another thing is efficiency: previously it was possible to have a fast version by setting the tolerance to 0 (i.e. no convergence test) but now the decomposition has become increasingly slow.

It might be good to have a review of where we're at, what features are actually needed and simplify the code a little.

**Comment by cohenjer** [comment]
+1. We will probably heavily tinker with the CP function with the dimension tree thing, so we should probably discuss how to proceed when we reach the implementation stage.

**Comment by cohenjer** [comment]
I ran a quick test just to provide more precise quantification of how much slower the function has become.

I decomposed ten 100x100x100 tensors using parafac with rank 10, default options except tol=0. This does not work for recent tensorly versions, so when the code fails I used return_errors=True (I believe this is a bug, will open a separate PR).
The total time is reported using time.perf_counter. Numpy backend.

I tried all tensorly version from the current dev to 0.3.0; here are the factual results:
`
Running latest tensorly version
Runtime for 10 CPs is 16.66701964100048
Running tensorly version 0.7.0
Runtime for 10 CPs is 14.145441272000426
Running tensorly version 0.6.0
Runtime for 10 CPs is 15.706550804000472
Running tensorly version 0.5.1
Runtime for 10 CPs is 13.24064584900043
Running tensorly version 0.5.0
Runtime for 10 CPs is 15.428937661000418
Running tensorly version 0.4.5
Runtime for 10 CPs is 13.758153636000316
Running tensorly version 0.4.4
Runtime for 10 CPs is 16.10983427699921
Running tensorly version 0.4.3
Runtime for 10 CPs is 5.178578480999931
Running tensorly version 0.4.2
Runtime for 10 CPs is 8.262633802999517
Running tensorly version 0.4.0
Runtime for 10 CPs is 9.182569708000301
Running tensorly version 0.3.0
Runtime for 10 CPs is 8.927185896999617
`

Notice how 0.4.3 is by far faster: this is because it is the only version where we actually do not compute the error when tol=0. For the other versions, it is not as clear for me why more recent ones are slower, I think the return_error flag being required is bad. It was needed for all version >= 0.4.5.  Might also be related to #442 but I did not investigate this seriously.

The bash script and python script are on [my github](https://github.com/cohenjer/testing_cp_speed_tensorly) if you want to cross check.

**Comment by JeanKossaifi** [comment]
Thanks, great to have numbers. I think the last profiling result is a little misleading, we see that the algo is significantly slower for newer version. It might be more informative to just look at the line-by-line profiling (`%lprun` in a notebook). 

I do think we should skip error calculations if we don't need (fixed number of iters) it as it slows things down quite a bit (that's why I had added option for `tol=0`).

**Comment by cohenjer** [comment]
Not sure I understand your comment on the profiling being misleading, the profiling is only for the last version anyway.

About error computation, after manual inspection it turns out we compute the error even if tol=0 in all versions>0.3 except:
- the latest one (main branch) !! (no I don't remember which PR changed this) (we still compute error if line search is on or if return_errors=True)
- version 0.4.3 
In a bunch of versions, tol=0 actually raises a bug because we don't want to compute the loss but the loss appears in some part of the code anyway.

So indeed I think 0.4.3 is faster than others because of error not being computed if tol=0, but it does not explain why the latest version is so bad (in fact it makes things even worse!)

**Comment by JeanKossaifi** [comment]
I agree with your points @cohenjer. As a side note, an easy way to speed mttkrp is to use the einsum tenalg backend. For the core one, as I mentioned in the issue you linked to, the change was made so TensorLy could be sparse safe. One solution is to redefine it in a sparse specific tensor backend. We also probably need to revisit how we handle sparse tensors, not sure how much the sparse backend is currently used.

Re profiling, the timings are great,  I was saying that line-profiling is probably more informative/easier to read than this report. 

For error computation, I fixed it in the commit I mentioned in the original comment. It's still needed in a few cases as you mentioned.

**Comment by cohenjer** [comment]
You are right @JeanKossaifi, in fact it is quite a bit of work to test all backends/tenalg-backend on all versions (but probably necessary to have a better picture). 

Also 100% agree with current sparse backend needing a lot of improvements.

About einsum backend: the thing is last time I checked einsum tenalg backend was horrible with numpy. 

Here are a few solutions to the *speed* issue of CP (cluttered code is a different story):
1. use einsum backend as default --> not sure how it would actually improve, in particular for numpy.
2. revert back to a faster default MTTKRP computation --> requires to move the current version to sparse backend in some way.

I would rather go with 2. but I will compute some tests when I find the time to see which is simpler/brings better improvements.

Any other ideas?

**Comment by JeanKossaifi** [comment]
I agree, I think 2 the way to go, ideally coordinated with some work on the sparse backend.

**Comment by cohenjer** [comment]
Just to update, I have not forgotten this issue; I am working on a quite large PR that will include this change (bot probably not coordinated with the sparse back-end)
